pandas/io/gbq.py

```diff
@@ -577,74 +540,251 @@ def generate_bq_schema(df, default_type='STRING'):
+            _Dataset(self.project_id).create(self.dataset_id)
+
+        body = {
+            'schema': schema,
+            'tableReference': {
+                'tableId': table_id,
+                'projectId': self.project_id,
+                'datasetId': self.dataset_id
+            }
+        }
+
+        try:
+            self.service.tables().insert(
+                 projectId=self.project_id,
+                 datasetId=self.dataset_id,
+                 body=body).execute()
+        except self.http_error as ex:
+            self.process_http_error(ex)
 
-    if table_exists(table, project_id):
-        raise TableCreationError("The table could not be created because it already exists")
+    def delete(self, table_id):
+        """ Delete a table in Google BigQuery
 
-    connector = GbqConnector(project_id)
-    dataset_id, table_id = table.rsplit('.', 1)
+        .. versionadded:: 0.17.0
 
-    return connector.create_table(dataset_id, table_id, schema)
+        Parameters
+        ----------
+        table : str
+            Name of table to be deleted
+        """
 
+        if not self.exists(table_id):
+            raise NotFoundException("Table does not exist")
 
-def delete_table(table, project_id):
-    """ Delete a table in Google BigQuery given a table and project id
+        try:
+            self.service.tables().delete(
+                 datasetId=self.dataset_id,
+                 projectId=self.project_id,
+                 tableId=table_id).execute()
+        except self.http_error as ex:
+            self.process_http_error(ex)
 
-    .. versionadded:: 0.17.0
 
-    Parameters
-    ----------
-    table : str
-        Name of table to be written, in the form 'dataset.tablename'
-    project_id : str
-        Google BigQuery Account project ID.
-    """
+class _Dataset(GbqConnector):
+
+    def __init__(self, project_id, reauth=False):
+        from apiclient.errors import HttpError
+        self.test_google_api_imports()
+        self.project_id = project_id
+        self.reauth = reauth
+        self.credentials = self.get_credentials()
+        self.service = self.get_service(self.credentials)
+        self.http_error = HttpError
+
+    def exists(self, dataset_id):
+        """ Check if a dataset exists in Google BigQuery
+
+        .. versionadded:: 0.17.0
+
+        Parameters
+        ----------
+        dataset_id : str
+            Name of dataset to be verified
+
+        Returns
+        -------
+        boolean
+            true if dataset exists, otherwise false
+        """
+
+        try:
+            self.service.datasets().get(
+                 projectId=self.project_id,
+                 datasetId=dataset_id).execute()
+            return True
+        except self.http_error as ex:
+            if ex.resp.status == 404:
+                return False
+            else:
+                self.process_http_error(ex)
+
+    def list(self):
```

Comment:
I would call this ``.datasets()`` rather than ``.list()``, or you could define ``__iter__`` which returns a list of the datasets (and ``__len__``) as well to make this a list like object.